### PR TITLE
W6 - AnhQuach - Canvas

### PR DESCRIPTION
**Pull request by at-anhquach**
#### Related issues (Optional)
Task #74 

#### Description (Required)
- [x] Draw Chart
- [x] Zoom in Zoom out
- [x] Scroll past

### CODE REVIEW

File: app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+    private List<Long> times = new ArrayList<>();
+
+    private float lefts[] = new float[mDistanceAs.size()];
+    private float prevXMove;
+    private float xDown;
+    private long timeDown;
+    private float v;
+    private float offsetX;
+    private boolean isMoveToRight;
+
```
Comment (at-tuannguyen) [inline app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java]: All variable members should be inited in the class's constructor, and start will `m` prefix

File: app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+    private Paint mPaintColumn2 = new Paint();
+    private Paint mPaintRect = new Paint();
+
+    private ScaleGestureDetector mScaleDetector;
+    private int columnWidth;
+    double distance;
+    private int mOxChart = getResources().getDimensionPixelSize(R.dimen.ox_chart);
+    private int columnCornerRadius = getResources().getDimensionPixelSize(R.dimen.column_corner_radius);
+    private int columnMarginHorizontal = getResources().getDimensionPixelSize(R.dimen.columns_margin_horizontal);
+    private float mWitdhScreen = ScreenUtil.getWidthScreen(getContext());
```
Comment (at-tuannguyen) [inline app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java]: -> `mWidthScreen`

File: app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+
+    private float startX = 0f;
+    private float startY = 0f;
+    private float translateX = 0f;
+    private float translateY = 0f;
+    private float previousTranslateX = 0f;
+    private float previousTranslateY = 0f;
+    private float mScaleFactor = 1.f;
+
+    private Paint mPaintColumn1 = new Paint();
```
Comment (at-tuannguyen) [inline app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java]: Please move it to line #44

File: app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+    private Paint mPaintText = new Paint();
+    private Paint mPaintLine = new Paint();
+    private Paint mPaintColumn = new Paint();
+    private Paint mPaintColumn2 = new Paint();
```
Comment (at-tuannguyen) [inline app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java]: The columns have the difference about its color, so, please set name to follow this format `mPaintColumn#Color`

File: app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+    private List<Integer> mDistanceCs = new ArrayList<>(Arrays.asList(3, 9, 5, 6, 10, 8, 9, 2, 4, 4, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9));
```
Comment (at-tuannguyen) [inline app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java]: Please rename follow format `mDistances#Color`

File: app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+    private Paint mPaintColumn1 = new Paint();
+    private Paint mPaintText = new Paint();
+    private Paint mPaintLine = new Paint();
+    private Paint mPaintColumn = new Paint();
+    private Paint mPaintColumn2 = new Paint();
+    private Paint mPaintRect = new Paint();
+
+    private ScaleGestureDetector mScaleDetector;
+    private int columnWidth;
+    double distance;
```
Comment (at-tuannguyen) [inline app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java]: This variable had assigned value, but don't use. Please remove it

File: app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+        mScaleDetector = new ScaleGestureDetector(context, new ScaleListener());
+        init();
+    }
+
+    private void init() {
+        mPaintText.setColor(getResources().getColor(R.color.colorGrayDark));
+        mPaintText.setTextSize(getResources().getDimension(R.dimen.textsize13sp));
+        mPaintText.setAntiAlias(true);
+
+        mPaintRect.setColor(getResources().getColor(R.color.colorWhite));
```
Comment (at-tuannguyen) [inline app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java]: With common color, you can use `Color.WHITE`, ...

File: app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+        mPaintColumn.setAntiAlias(true);
+
+        mPaintColumn1.setColor(getResources().getColor(R.color.colorCyanA700));
+        mPaintColumn1.setAntiAlias(true);
+
+        mPaintColumn2.setColor(getResources().getColor(R.color.colorOrange500));
+        mPaintColumn2.setAntiAlias(true);
+    }
+
+    private float maxLists() {
```
Comment (at-tuannguyen) [inline app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java]: By now, the data is not changing, so please get max value after `init()`, it mean you need to define a variable hold this value

File: app/src/at-anhquach/res/values/attrs.xml [hunk 9]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <declare-styleable name="ChartView">
+        <attr name="column_width" format="integer" />
```
Comment (at-tuannguyen) [inline app/src/at-anhquach/res/values/attrs.xml]: `format ` should be `dimention`

File: app/src/at-anhquach/res/values/attrs.xml [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    </declare-styleable>
+
+<declare-styleable name="InteractiveLineGraphView">
```
Comment (at-tuannguyen) [inline app/src/at-anhquach/res/values/attrs.xml]: This `attr` doesn't use anywhere, so please remove it

File: app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+        drawText(canvas, getContext().getString(R.string.distance_km, ((int) maxLists())), 0, maxLists(), mPaintText);
+        drawText(canvas, getContext().getString(R.string.day), mOxChart, -1, mPaintText);
+        drawText(canvas, getContext().getString(R.string.month), mWitdhScreen - 100, -1, mPaintText);
+
+        drawLine(canvas, 0, mPaintLine);
+        drawLine(canvas, maxLists(), mPaintLine);
+
+        drawText(canvas, getContext().getString(R.string.distance_km, ((int) 0)), 0, 0, mPaintText);
+
+        drawLine(canvas, maxLists() / 2, mPaintLine);
```
Comment (at-tuannguyen) [inline app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java]: Draw line here make the line place above the chart, please fix it

File: app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+                break;
+        }
+        mScaleDetector.onTouchEvent(ev);
+        if ((mode == DRAG && mScaleFactor != 1f && dragged) || mode == ZOOM) {
+            invalidate();
+        }
+        return true;
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
```
Comment (at-tuannguyen) [inline app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java]: By now, it's ok. But in project, must be fix it

File: app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+
+        drawLine(canvas, maxLists() / 2, mPaintLine);
+        canvas.restore();
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void drawChart(Canvas canvas) {
+        float leftRect = getWidth() - getPaddingRight() - columnMarginHorizontal - columnWidth - 100 - offsetX;
+        for (int index = 0; index < mDistanceAs.size(); index++) {
+            canvas.drawRoundRect(leftRect - 100 * index, getTop(mDistanceAs.get(index)), leftRect - 100 * index + columnWidth,
```
Comment (at-tuannguyen) [inline app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java]: `100` use 14 times, please define a constant hold this value

File: app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java (hunk 11 above)
Comment (at-anhquach) [reply app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java]: I don't understand this comment, please explain for me, thank you!

File: app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+    private List<Integer> mDistanceOrange500s = new ArrayList<>(Arrays.asList(3, 9, 5, 6, 10, 8, 9, 2, 4, 4, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9, 2, 4, 5, 6, 7, 8, 9));
+    private List<Float> mPaths = new ArrayList<>();
+    private List<Long> mTimes = new ArrayList<>();
+
+    private float mLefts[] = new float[mDistancePurple800s.size()];
```
Comment (at-tuannguyen) [inline app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java]: All variable members should be inited in the class's constructor

File: app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java [hunk 15]
```diff
@@ -0,0 +1,341 @@
...
+    }
+
+    @SuppressLint("ClickableViewAccessibility")
+    public ChartView(Context context, @Nullable AttributeSet attrs) {
+        super(context, attrs);
+        TypedArray a = context.getTheme().obtainStyledAttributes(
+                attrs, R.styleable.ChartView, 0, 0
+        );
+        try {
+            mColumnWidth = (int) a.getDimension(R.styleable.ChartView_column_width, 10);
```
Comment (at-tuannguyen) [inline app/src/at-anhquach/java/vn/asiantech/internship/ui/canvas/ChartView.java]: Use `getDimensionPixelSize` instead, `10` should be get from `dimen.xml`

**Pull request merged**